In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

#imported the Necessary Library 

In [2]:
#reading the Data

df=pd.read_csv(r'C:\\Users\\HP\Downloads\transactions (1).csv')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106020 entries, 0 to 106019
Data columns (total 10 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   tid          106020 non-null  object 
 1   datetime     106020 non-null  object 
 2   cc_num       106020 non-null  int64  
 3   category     106020 non-null  object 
 4   amount       106020 non-null  float64
 5   latitude     106020 non-null  float64
 6   longitude    106020 non-null  float64
 7   city         106020 non-null  object 
 8   country      106020 non-null  object 
 9   fraud_label  106020 non-null  int64  
dtypes: float64(3), int64(2), object(5)
memory usage: 8.1+ MB


In [3]:
#reading files


fd=pd.read_csv(r'C:\\Users\\HP\Downloads\profiles.csv')

fd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   name       2000 non-null   object
 1   sex        2000 non-null   object
 2   mail       2000 non-null   object
 3   birthdate  2000 non-null   object
 4   City       2000 non-null   object
 5   Country    2000 non-null   object
 6   cc_num     2000 non-null   int64 
dtypes: int64(1), object(6)
memory usage: 109.5+ KB


In [4]:
#redaing The files


dframe=pd.read_csv(r'C:\\Users\\HP\Downloads\credit_cards_with_date_added.csv')

dframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  2000 non-null   int64 
 1   cc_num      2000 non-null   int64 
 2   provider    2000 non-null   object
 3   expires     2000 non-null   object
 4   date_added  2000 non-null   object
dtypes: int64(2), object(3)
memory usage: 78.2+ KB


In [5]:
#reading the ffiles

dff=pd.read_csv('C:\\Users\\HP\Downloads\credit_cards_with_date_added.csv')

dff.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  2000 non-null   int64 
 1   cc_num      2000 non-null   int64 
 2   provider    2000 non-null   object
 3   expires     2000 non-null   object
 4   date_added  2000 non-null   object
dtypes: int64(2), object(3)
memory usage: 78.2+ KB


We gat 4 dataset,dff and fd including dframe have the same data in tem of rows and columns.

So what Next we will do is we joining them together.

In [6]:
#viw the data for dff

dff.head(2)

Unnamed: 0            cc_num provider expires        date_added
0           0  4796807885357879     visa   05/23  2022-09-01 19:50
1           1  4529266636192966     visa   03/22  2022-09-01 19:50

In [7]:
#vieing the Data for dframe

dframe.head(2)

Unnamed: 0            cc_num provider expires        date_added
0           0  4796807885357879     visa   05/23  2022-09-01 19:50
1           1  4529266636192966     visa   03/22  2022-09-01 19:50

looking this Two dataset i found out that both dff and dframe are made up of the same dataset so I have to used one

In [8]:
fd.head(2)


#seems to Look like a complete diffferent data from dframe and fd
#but one thing is the dframe,fd, and dff and the same Length of data, of fd columns seems to look different from dff
#am joining fd with dframe, and ThankGodd they have the on eunique columns Name cc_num

name sex                       mail   birthdate  \
0  Catherine Zimmerman   F  valenciajason@hotmail.com  1988-09-20   
1     Michael Williams   M     brettkennedy@yahoo.com  1977-03-01   

                City Country            cc_num  
0   Bryn Mawr-Skyway      US  4796807885357879  
1  Gates-North Gates      US  4529266636192966

In [9]:
#create a New dataframe y joining dframe and fd

ndf=dframe.merge(fd,on='cc_num')

In [10]:


ndf

Unnamed: 0            cc_num provider expires        date_added  \
0              0  4796807885357879     visa   05/23  2022-09-01 19:50   
1              1  4529266636192966     visa   03/22  2022-09-01 19:50   
2              2  4922690008243953     visa   02/27  2022-09-01 19:50   
3              3  4897369589533543     visa   04/22  2022-09-01 19:50   
4              4  4848518335893425     visa   10/26  2022-09-01 19:50   
...          ...               ...      ...     ...               ...   
1995        1995  4473593503484549     visa   01/26  2022-09-01 19:50   
1996        1996  4483607926610254     visa   08/26  2022-09-01 19:50   
1997        1997  4249540384047310     visa   10/22  2022-09-01 19:50   
1998        1998  4325889817240225     visa   03/24  2022-09-01 19:50   
1999        1999  4653925648407310     visa   04/22  2022-09-01 19:50   

                     name sex                       mail   birthdate  \
0     Catherine Zimmerman   F  valenciajason@hotmail.com  1988-09-20   
1        Michael Williams   M     brettkennedy@yahoo.com  1977-03-01   
2         Jessica Krueger   F     marthacruz@hotmail.com  1947-09-10   
3             Ruth Harris   F          james11@yahoo.com  1983-12-27   
4             Paul Ashley   M      matthew97@hotmail.com  1974-11-10   
...                   ...  ..                        ...         ...   
1995        Megan Ramirez   F        jmorrison@gmail.com  1924-06-28   
1996        Jessica Gould   F        larrydiaz@yahoo.com  1925-11-07   
1997           Adam Black   M     richardrusso@yahoo.com  1980-09-15   
1998      Brooke Robinson   F        cwagner@hotmail.com  1986-11-05   
1999      Mr. Jeff Torres   M         robert76@gmail.com  1932-11-13   

                   City Country  
0      Bryn Mawr-Skyway      US  
1     Gates-North Gates      US  
2            Greenfield      US  
3              New City      US  
4               Peabody      US  
...                 ...     ...  
1995        Stockbridge      US  
1996      Port Richmond      US  
1997  Brentwood Estates      US  
1998            Redding      US  
1999             Tupelo      US  

[2000 rows x 11 columns]

In [11]:
#checking this data, but is totally diiferent 

df.head()

tid             datetime            cc_num  \
0  11df919988c134d97bbff2678eb68e22  2022-01-01 00:00:24  4473593503484549   
1  dd0b2d6d4266ccd3bf05bc2ea91cf180  2022-01-01 00:00:56  4272465718946864   
2  e627f5d9a9739833bd52d2da51761fc3  2022-01-01 00:02:32  4104216579248948   
3  6fb3e6beafbb92b8e15827037f603c52  2022-01-01 00:03:24  4814447237003448   
4  be0b8acc57bfe126a5a392fd99e6ddd1  2022-01-01 00:03:55  4515188652242507   

             category  amount  latitude  longitude            city country  \
0       Health/Beauty   62.95  42.30865  -83.48216          Canton      US   
1             Grocery   85.45  33.52253 -117.70755   Laguna Niguel      US   
2  Domestic Transport   21.63  37.60876  -77.37331  Mechanicsville      US   
3       Health/Beauty   54.71  43.54072 -116.56346           Nampa      US   
4             Grocery   59.22  40.24537  -75.64963       Pottstown      US   

   fraud_label  
0            0  
1            0  
2            0  
3            0  
4            0

In [12]:
#checking the datainfo

ndf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2000 entries, 0 to 1999
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  2000 non-null   int64 
 1   cc_num      2000 non-null   int64 
 2   provider    2000 non-null   object
 3   expires     2000 non-null   object
 4   date_added  2000 non-null   object
 5   name        2000 non-null   object
 6   sex         2000 non-null   object
 7   mail        2000 non-null   object
 8   birthdate   2000 non-null   object
 9   City        2000 non-null   object
 10  Country     2000 non-null   object
dtypes: int64(2), object(9)
memory usage: 187.5+ KB


In [13]:
#dropping unncessarycolumns

ndf=ndf.drop('Unnamed: 0',axis=1)

ndf

cc_num provider expires        date_added  \
0     4796807885357879     visa   05/23  2022-09-01 19:50   
1     4529266636192966     visa   03/22  2022-09-01 19:50   
2     4922690008243953     visa   02/27  2022-09-01 19:50   
3     4897369589533543     visa   04/22  2022-09-01 19:50   
4     4848518335893425     visa   10/26  2022-09-01 19:50   
...                ...      ...     ...               ...   
1995  4473593503484549     visa   01/26  2022-09-01 19:50   
1996  4483607926610254     visa   08/26  2022-09-01 19:50   
1997  4249540384047310     visa   10/22  2022-09-01 19:50   
1998  4325889817240225     visa   03/24  2022-09-01 19:50   
1999  4653925648407310     visa   04/22  2022-09-01 19:50   

                     name sex                       mail   birthdate  \
0     Catherine Zimmerman   F  valenciajason@hotmail.com  1988-09-20   
1        Michael Williams   M     brettkennedy@yahoo.com  1977-03-01   
2         Jessica Krueger   F     marthacruz@hotmail.com  1947-09-10   
3             Ruth Harris   F          james11@yahoo.com  1983-12-27   
4             Paul Ashley   M      matthew97@hotmail.com  1974-11-10   
...                   ...  ..                        ...         ...   
1995        Megan Ramirez   F        jmorrison@gmail.com  1924-06-28   
1996        Jessica Gould   F        larrydiaz@yahoo.com  1925-11-07   
1997           Adam Black   M     richardrusso@yahoo.com  1980-09-15   
1998      Brooke Robinson   F        cwagner@hotmail.com  1986-11-05   
1999      Mr. Jeff Torres   M         robert76@gmail.com  1932-11-13   

                   City Country  
0      Bryn Mawr-Skyway      US  
1     Gates-North Gates      US  
2            Greenfield      US  
3              New City      US  
4               Peabody      US  
...                 ...     ...  
1995        Stockbridge      US  
1996      Port Richmond      US  
1997  Brentwood Estates      US  
1998            Redding      US  
1999             Tupelo      US  

[2000 rows x 10 columns]

In [14]:
#getting the date columns to prooper format


ndf['birthdate']=pd.to_datetime(ndf['birthdate'],errors='coerce')

In [15]:
ndf

cc_num provider expires        date_added  \
0     4796807885357879     visa   05/23  2022-09-01 19:50   
1     4529266636192966     visa   03/22  2022-09-01 19:50   
2     4922690008243953     visa   02/27  2022-09-01 19:50   
3     4897369589533543     visa   04/22  2022-09-01 19:50   
4     4848518335893425     visa   10/26  2022-09-01 19:50   
...                ...      ...     ...               ...   
1995  4473593503484549     visa   01/26  2022-09-01 19:50   
1996  4483607926610254     visa   08/26  2022-09-01 19:50   
1997  4249540384047310     visa   10/22  2022-09-01 19:50   
1998  4325889817240225     visa   03/24  2022-09-01 19:50   
1999  4653925648407310     visa   04/22  2022-09-01 19:50   

                     name sex                       mail  birthdate  \
0     Catherine Zimmerman   F  valenciajason@hotmail.com 1988-09-20   
1        Michael Williams   M     brettkennedy@yahoo.com 1977-03-01   
2         Jessica Krueger   F     marthacruz@hotmail.com 1947-09-10   
3             Ruth Harris   F          james11@yahoo.com 1983-12-27   
4             Paul Ashley   M      matthew97@hotmail.com 1974-11-10   
...                   ...  ..                        ...        ...   
1995        Megan Ramirez   F        jmorrison@gmail.com 1924-06-28   
1996        Jessica Gould   F        larrydiaz@yahoo.com 1925-11-07   
1997           Adam Black   M     richardrusso@yahoo.com 1980-09-15   
1998      Brooke Robinson   F        cwagner@hotmail.com 1986-11-05   
1999      Mr. Jeff Torres   M         robert76@gmail.com 1932-11-13   

                   City Country  
0      Bryn Mawr-Skyway      US  
1     Gates-North Gates      US  
2            Greenfield      US  
3              New City      US  
4               Peabody      US  
...                 ...     ...  
1995        Stockbridge      US  
1996      Port Richmond      US  
1997  Brentwood Estates      US  
1998            Redding      US  
1999             Tupelo      US  

[2000 rows x 10 columns]

In [16]:
#Splitting the Date columns 


ndf[['Date','Time']]=ndf['date_added'].str.split(' ',expand=True)

In [17]:
ndf

cc_num provider expires        date_added  \
0     4796807885357879     visa   05/23  2022-09-01 19:50   
1     4529266636192966     visa   03/22  2022-09-01 19:50   
2     4922690008243953     visa   02/27  2022-09-01 19:50   
3     4897369589533543     visa   04/22  2022-09-01 19:50   
4     4848518335893425     visa   10/26  2022-09-01 19:50   
...                ...      ...     ...               ...   
1995  4473593503484549     visa   01/26  2022-09-01 19:50   
1996  4483607926610254     visa   08/26  2022-09-01 19:50   
1997  4249540384047310     visa   10/22  2022-09-01 19:50   
1998  4325889817240225     visa   03/24  2022-09-01 19:50   
1999  4653925648407310     visa   04/22  2022-09-01 19:50   

                     name sex                       mail  birthdate  \
0     Catherine Zimmerman   F  valenciajason@hotmail.com 1988-09-20   
1        Michael Williams   M     brettkennedy@yahoo.com 1977-03-01   
2         Jessica Krueger   F     marthacruz@hotmail.com 1947-09-10   
3             Ruth Harris   F          james11@yahoo.com 1983-12-27   
4             Paul Ashley   M      matthew97@hotmail.com 1974-11-10   
...                   ...  ..                        ...        ...   
1995        Megan Ramirez   F        jmorrison@gmail.com 1924-06-28   
1996        Jessica Gould   F        larrydiaz@yahoo.com 1925-11-07   
1997           Adam Black   M     richardrusso@yahoo.com 1980-09-15   
1998      Brooke Robinson   F        cwagner@hotmail.com 1986-11-05   
1999      Mr. Jeff Torres   M         robert76@gmail.com 1932-11-13   

                   City Country        Date   Time  
0      Bryn Mawr-Skyway      US  2022-09-01  19:50  
1     Gates-North Gates      US  2022-09-01  19:50  
2            Greenfield      US  2022-09-01  19:50  
3              New City      US  2022-09-01  19:50  
4               Peabody      US  2022-09-01  19:50  
...                 ...     ...         ...    ...  
1995        Stockbridge      US  2022-09-01  19:50  
1996      Port Richmond      US  2022-09-01  19:50  
1997  Brentwood Estates      US  2022-09-01  19:50  
1998            Redding      US  2022-09-01  19:50  
1999             Tupelo      US  2022-09-01  19:50  

[2000 rows x 12 columns]

In [18]:
#droping The Date_added columns

ndf=ndf.drop('date_added',axis=1)

In [19]:
#then we rename the columns

ndf=ndf.rename(columns={'Date':'AddedDate','Time':'AddTime'})

In [20]:
ndf

cc_num provider expires                 name sex  \
0     4796807885357879     visa   05/23  Catherine Zimmerman   F   
1     4529266636192966     visa   03/22     Michael Williams   M   
2     4922690008243953     visa   02/27      Jessica Krueger   F   
3     4897369589533543     visa   04/22          Ruth Harris   F   
4     4848518335893425     visa   10/26          Paul Ashley   M   
...                ...      ...     ...                  ...  ..   
1995  4473593503484549     visa   01/26        Megan Ramirez   F   
1996  4483607926610254     visa   08/26        Jessica Gould   F   
1997  4249540384047310     visa   10/22           Adam Black   M   
1998  4325889817240225     visa   03/24      Brooke Robinson   F   
1999  4653925648407310     visa   04/22      Mr. Jeff Torres   M   

                           mail  birthdate               City Country  \
0     valenciajason@hotmail.com 1988-09-20   Bryn Mawr-Skyway      US   
1        brettkennedy@yahoo.com 1977-03-01  Gates-North Gates      US   
2        marthacruz@hotmail.com 1947-09-10         Greenfield      US   
3             james11@yahoo.com 1983-12-27           New City      US   
4         matthew97@hotmail.com 1974-11-10            Peabody      US   
...                         ...        ...                ...     ...   
1995        jmorrison@gmail.com 1924-06-28        Stockbridge      US   
1996        larrydiaz@yahoo.com 1925-11-07      Port Richmond      US   
1997     richardrusso@yahoo.com 1980-09-15  Brentwood Estates      US   
1998        cwagner@hotmail.com 1986-11-05            Redding      US   
1999         robert76@gmail.com 1932-11-13             Tupelo      US   

       AddedDate AddTime  
0     2022-09-01   19:50  
1     2022-09-01   19:50  
2     2022-09-01   19:50  
3     2022-09-01   19:50  
4     2022-09-01   19:50  
...          ...     ...  
1995  2022-09-01   19:50  
1996  2022-09-01   19:50  
1997  2022-09-01   19:50  
1998  2022-09-01   19:50  
1999  2022-09-01   19:50  

[2000 rows x 11 columns]

In [21]:
ndf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2000 entries, 0 to 1999
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   cc_num     2000 non-null   int64         
 1   provider   2000 non-null   object        
 2   expires    2000 non-null   object        
 3   name       2000 non-null   object        
 4   sex        2000 non-null   object        
 5   mail       2000 non-null   object        
 6   birthdate  2000 non-null   datetime64[ns]
 7   City       2000 non-null   object        
 8   Country    2000 non-null   object        
 9   AddedDate  2000 non-null   object        
 10  AddTime    2000 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(9)
memory usage: 187.5+ KB


In [22]:
#Getting AddedDate into proper format

ndf['AddedDate']=pd.to_datetime(ndf['AddedDate'],errors='coerce')

In [23]:
ndf

cc_num provider expires                 name sex  \
0     4796807885357879     visa   05/23  Catherine Zimmerman   F   
1     4529266636192966     visa   03/22     Michael Williams   M   
2     4922690008243953     visa   02/27      Jessica Krueger   F   
3     4897369589533543     visa   04/22          Ruth Harris   F   
4     4848518335893425     visa   10/26          Paul Ashley   M   
...                ...      ...     ...                  ...  ..   
1995  4473593503484549     visa   01/26        Megan Ramirez   F   
1996  4483607926610254     visa   08/26        Jessica Gould   F   
1997  4249540384047310     visa   10/22           Adam Black   M   
1998  4325889817240225     visa   03/24      Brooke Robinson   F   
1999  4653925648407310     visa   04/22      Mr. Jeff Torres   M   

                           mail  birthdate               City Country  \
0     valenciajason@hotmail.com 1988-09-20   Bryn Mawr-Skyway      US   
1        brettkennedy@yahoo.com 1977-03-01  Gates-North Gates      US   
2        marthacruz@hotmail.com 1947-09-10         Greenfield      US   
3             james11@yahoo.com 1983-12-27           New City      US   
4         matthew97@hotmail.com 1974-11-10            Peabody      US   
...                         ...        ...                ...     ...   
1995        jmorrison@gmail.com 1924-06-28        Stockbridge      US   
1996        larrydiaz@yahoo.com 1925-11-07      Port Richmond      US   
1997     richardrusso@yahoo.com 1980-09-15  Brentwood Estates      US   
1998        cwagner@hotmail.com 1986-11-05            Redding      US   
1999         robert76@gmail.com 1932-11-13             Tupelo      US   

      AddedDate AddTime  
0    2022-09-01   19:50  
1    2022-09-01   19:50  
2    2022-09-01   19:50  
3    2022-09-01   19:50  
4    2022-09-01   19:50  
...         ...     ...  
1995 2022-09-01   19:50  
1996 2022-09-01   19:50  
1997 2022-09-01   19:50  
1998 2022-09-01   19:50  
1999 2022-09-01   19:50  

[2000 rows x 11 columns]

In [24]:
#checking out the Sex

ndf['sex'].value_counts()

M    1010
F     990
Name: sex, dtype: int64

In [25]:
#checking out the City

ndf['City'].value_counts()

Newburgh          24
Estelle           20
Ada               19
Lino Lakes        18
Malden            18
                  ..
South Whittier     5
Nampa              5
Bon Air            5
Beaumont           4
Middle River       3
Name: City, Length: 161, dtype: int64

In [26]:
#checking for the max date

ndf['birthdate'].max()

Timestamp('2004-04-09 00:00:00')

In [27]:
#checking for the min

ndf['birthdate'].min()

Timestamp('1921-07-03 00:00:00')

In [28]:
#query out foe the oldest man in the building

ndf.query('birthdate=="1921-07-03"')

cc_num provider expires           name sex  \
1602  4983841885614867     visa   03/22  Erica Rowland   F   

                           mail  birthdate    City Country  AddedDate AddTime  
1602  jameshutchinson@gmail.com 1921-07-03  Dumont      US 2022-09-01   19:50

In [29]:
#Querying out Youngest man


ndf.query('birthdate=="2004-04-09"')

cc_num provider expires          name sex  \
1505  4565244478713673     visa   04/23  Jason Morris   M   

                        mail  birthdate           City Country  AddedDate  \
1505  jamiebrown@hotmail.com 2004-04-09  Scotch Plains      US 2022-09-01   

     AddTime  
1505   19:50

In [30]:
#checking for the latest date


ndf['expires'].min() 

'01/22'

In [31]:
#checkinf ro the farest date

ndf['expires'].max()

'12/26'

In [32]:
#importing the date to help with today date

from datetime import datetime

Data=datetime.now()

In [33]:
#date nature

ndf['expires'].unique()

array(['05/23', '03/22', '02/27', '04/22', '10/26', '02/24', '08/24',
       '11/26', '05/25', '05/27', '05/24', '02/23', '09/24', '11/25',
       '07/25', '11/22', '01/24', '12/26', '03/23', '02/22', '11/23',
       '09/23', '04/23', '03/24', '09/22', '12/22', '08/23', '10/24',
       '08/22', '03/27', '09/25', '01/25', '12/23', '05/26', '12/25',
       '10/23', '10/25', '03/25', '07/24', '09/26', '07/26', '02/25',
       '10/22', '05/22', '01/26', '04/26', '06/22', '02/26', '01/22',
       '01/27', '08/25', '04/27', '08/26', '01/23', '07/22', '11/24',
       '06/26', '06/25', '07/23', '03/26', '12/24', '04/24', '04/25',
       '06/23', '06/24'], dtype=object)

In [34]:
#confirming the date

ndf['Today_Date']=Data

In [35]:
ndf

cc_num provider expires                 name sex  \
0     4796807885357879     visa   05/23  Catherine Zimmerman   F   
1     4529266636192966     visa   03/22     Michael Williams   M   
2     4922690008243953     visa   02/27      Jessica Krueger   F   
3     4897369589533543     visa   04/22          Ruth Harris   F   
4     4848518335893425     visa   10/26          Paul Ashley   M   
...                ...      ...     ...                  ...  ..   
1995  4473593503484549     visa   01/26        Megan Ramirez   F   
1996  4483607926610254     visa   08/26        Jessica Gould   F   
1997  4249540384047310     visa   10/22           Adam Black   M   
1998  4325889817240225     visa   03/24      Brooke Robinson   F   
1999  4653925648407310     visa   04/22      Mr. Jeff Torres   M   

                           mail  birthdate               City Country  \
0     valenciajason@hotmail.com 1988-09-20   Bryn Mawr-Skyway      US   
1        brettkennedy@yahoo.com 1977-03-01  Gates-North Gates      US   
2        marthacruz@hotmail.com 1947-09-10         Greenfield      US   
3             james11@yahoo.com 1983-12-27           New City      US   
4         matthew97@hotmail.com 1974-11-10            Peabody      US   
...                         ...        ...                ...     ...   
1995        jmorrison@gmail.com 1924-06-28        Stockbridge      US   
1996        larrydiaz@yahoo.com 1925-11-07      Port Richmond      US   
1997     richardrusso@yahoo.com 1980-09-15  Brentwood Estates      US   
1998        cwagner@hotmail.com 1986-11-05            Redding      US   
1999         robert76@gmail.com 1932-11-13             Tupelo      US   

      AddedDate AddTime                 Today_Date  
0    2022-09-01   19:50 2023-10-04 06:28:40.248442  
1    2022-09-01   19:50 2023-10-04 06:28:40.248442  
2    2022-09-01   19:50 2023-10-04 06:28:40.248442  
3    2022-09-01   19:50 2023-10-04 06:28:40.248442  
4    2022-09-01   19:50 2023-10-04 06:28:40.248442  
...         ...     ...                        ...  
1995 2022-09-01   19:50 2023-10-04 06:28:40.248442  
1996 2022-09-01   19:50 2023-10-04 06:28:40.248442  
1997 2022-09-01   19:50 2023-10-04 06:28:40.248442  
1998 2022-09-01   19:50 2023-10-04 06:28:40.248442  
1999 2022-09-01   19:50 2023-10-04 06:28:40.248442  

[2000 rows x 12 columns]

In [36]:
#checking the date data type

type(ndf['Today_Date'][0])

pandas._libs.tslibs.timestamps.Timestamp

In [37]:
#also for the Expire date columns

type(ndf['expires'][0])

str

In [46]:
ndf['expires']=ndf['expires'].str.replace('/','-',regex=True)

In [48]:
ndf['expires']=pd.to_datetime(ndf['expires'],format='%m-%y')

In [50]:
ndf

cc_num provider    expires                 name sex  \
0     4796807885357879     visa 2023-05-01  Catherine Zimmerman   F   
1     4529266636192966     visa 2022-03-01     Michael Williams   M   
2     4922690008243953     visa 2027-02-01      Jessica Krueger   F   
3     4897369589533543     visa 2022-04-01          Ruth Harris   F   
4     4848518335893425     visa 2026-10-01          Paul Ashley   M   
...                ...      ...        ...                  ...  ..   
1995  4473593503484549     visa 2026-01-01        Megan Ramirez   F   
1996  4483607926610254     visa 2026-08-01        Jessica Gould   F   
1997  4249540384047310     visa 2022-10-01           Adam Black   M   
1998  4325889817240225     visa 2024-03-01      Brooke Robinson   F   
1999  4653925648407310     visa 2022-04-01      Mr. Jeff Torres   M   

                           mail  birthdate               City Country  \
0     valenciajason@hotmail.com 1988-09-20   Bryn Mawr-Skyway      US   
1        brettkennedy@yahoo.com 1977-03-01  Gates-North Gates      US   
2        marthacruz@hotmail.com 1947-09-10         Greenfield      US   
3             james11@yahoo.com 1983-12-27           New City      US   
4         matthew97@hotmail.com 1974-11-10            Peabody      US   
...                         ...        ...                ...     ...   
1995        jmorrison@gmail.com 1924-06-28        Stockbridge      US   
1996        larrydiaz@yahoo.com 1925-11-07      Port Richmond      US   
1997     richardrusso@yahoo.com 1980-09-15  Brentwood Estates      US   
1998        cwagner@hotmail.com 1986-11-05            Redding      US   
1999         robert76@gmail.com 1932-11-13             Tupelo      US   

      AddedDate AddTime                 Today_Date  
0    2022-09-01   19:50 2023-10-04 06:28:40.248442  
1    2022-09-01   19:50 2023-10-04 06:28:40.248442  
2    2022-09-01   19:50 2023-10-04 06:28:40.248442  
3    2022-09-01   19:50 2023-10-04 06:28:40.248442  
4    2022-09-01   19:50 2023-10-04 06:28:40.248442  
...         ...     ...                        ...  
1995 2022-09-01   19:50 2023-10-04 06:28:40.248442  
1996 2022-09-01   19:50 2023-10-04 06:28:40.248442  
1997 2022-09-01   19:50 2023-10-04 06:28:40.248442  
1998 2022-09-01   19:50 2023-10-04 06:28:40.248442  
1999 2022-09-01   19:50 2023-10-04 06:28:40.248442  

[2000 rows x 12 columns]

In [55]:
ndf['Today_Date']

0      2023-10-04 06:28:40.248442
1      2023-10-04 06:28:40.248442
2      2023-10-04 06:28:40.248442
3      2023-10-04 06:28:40.248442
4      2023-10-04 06:28:40.248442
                  ...            
1995   2023-10-04 06:28:40.248442
1996   2023-10-04 06:28:40.248442
1997   2023-10-04 06:28:40.248442
1998   2023-10-04 06:28:40.248442
1999   2023-10-04 06:28:40.248442
Name: Today_Date, Length: 2000, dtype: datetime64[ns]

In [60]:
type(ndf['Today_Date'][0])

pandas._libs.tslibs.timestamps.Timestamp

In [62]:
ndf['Today_Date']=pd.to_datetime(ndf['Today_Date'])

In [64]:
#let check for the Their age


ndf['Age']=pd.DatetimeIndex(ndf['Today_Date']).year -pd.DatetimeIndex(ndf['birthdate']).year

In [65]:
ndf.head()

cc_num provider    expires                 name sex  \
0  4796807885357879     visa 2023-05-01  Catherine Zimmerman   F   
1  4529266636192966     visa 2022-03-01     Michael Williams   M   
2  4922690008243953     visa 2027-02-01      Jessica Krueger   F   
3  4897369589533543     visa 2022-04-01          Ruth Harris   F   
4  4848518335893425     visa 2026-10-01          Paul Ashley   M   

                        mail  birthdate               City Country  AddedDate  \
0  valenciajason@hotmail.com 1988-09-20   Bryn Mawr-Skyway      US 2022-09-01   
1     brettkennedy@yahoo.com 1977-03-01  Gates-North Gates      US 2022-09-01   
2     marthacruz@hotmail.com 1947-09-10         Greenfield      US 2022-09-01   
3          james11@yahoo.com 1983-12-27           New City      US 2022-09-01   
4      matthew97@hotmail.com 1974-11-10            Peabody      US 2022-09-01   

  AddTime                 Today_Date  Age  
0   19:50 2023-10-04 06:28:40.248442   35  
1   19:50 2023-10-04 06:28:40.248442   46  
2   19:50 2023-10-04 06:28:40.248442   76  
3   19:50 2023-10-04 06:28:40.248442   40  
4   19:50 2023-10-04 06:28:40.248442   49

In [66]:
#checking for the reminding days or whose card as expires?

ndf["days_until_card_expires"] = (ndf["expires"]-ndf["Today_Date"]) / np.timedelta64(1, "D")

In [67]:
ndf.head()

cc_num provider    expires                 name sex  \
0  4796807885357879     visa 2023-05-01  Catherine Zimmerman   F   
1  4529266636192966     visa 2022-03-01     Michael Williams   M   
2  4922690008243953     visa 2027-02-01      Jessica Krueger   F   
3  4897369589533543     visa 2022-04-01          Ruth Harris   F   
4  4848518335893425     visa 2026-10-01          Paul Ashley   M   

                        mail  birthdate               City Country  AddedDate  \
0  valenciajason@hotmail.com 1988-09-20   Bryn Mawr-Skyway      US 2022-09-01   
1     brettkennedy@yahoo.com 1977-03-01  Gates-North Gates      US 2022-09-01   
2     marthacruz@hotmail.com 1947-09-10         Greenfield      US 2022-09-01   
3          james11@yahoo.com 1983-12-27           New City      US 2022-09-01   
4      matthew97@hotmail.com 1974-11-10            Peabody      US 2022-09-01   

  AddTime                 Today_Date  Age  days_until_card_expires  
0   19:50 2023-10-04 06:28:40.248442   35               -156.26991  
1   19:50 2023-10-04 06:28:40.248442   46               -582.26991  
2   19:50 2023-10-04 06:28:40.248442   76               1215.73009  
3   19:50 2023-10-04 06:28:40.248442   40               -551.26991  
4   19:50 2023-10-04 06:28:40.248442   49               1092.73009

NOte :
    
    the Custumers Whose  card gat minus sign(-) means the date of his/her card has expires why the onces whose card is
    doesn't or has no negative sign in front of it mean it till valid.

In [74]:
#let run the maximum and the minimum:


def minmax(X):
    mini=ndf[X].idxmin()
    maxi=ndf[X].idxmax()
    maxii=pd.DataFrame(ndf.loc[maxi,:])
    minii=pd.DataFrame(ndf.loc[mini,:])
    return pd.concat([maxii,minii],axis=1)

In [75]:
minmax('days_until_card_expires')

9   \
cc_num                             4577917577528649   
provider                                       visa   
expires                         2027-05-01 00:00:00   
name                                 David Stephens   
sex                                               M   
mail                          moodylauren@gmail.com   
birthdate                       1933-10-29 00:00:00   
City                                      Pottstown   
Country                                          US   
AddedDate                       2022-09-01 00:00:00   
AddTime                                       19:50   
Today_Date               2023-10-04 06:28:40.248442   
Age                                              90   
days_until_card_expires                  1304.73009   
age_at_transaction                         89.93277   

                                                 76  
cc_num                             4996574944074667  
provider                                       visa  
expires                         2022-01-01 00:00:00  
name                                   John Sanders  
sex                                               M  
mail                           millerpaul@gmail.com  
birthdate                       1976-04-13 00:00:00  
City                                        Waverly  
Country                                          US  
AddedDate                       2022-09-01 00:00:00  
AddTime                                       19:50  
Today_Date               2023-10-04 06:28:40.248442  
Age                                              47  
days_until_card_expires                  -641.26991  
age_at_transaction                        47.476046

Data Observation:

    so we gat The Name David Stephnes whose card will card on 2027 of 05 
    
    and we gat John Sanders whose card expire years back. and her age is about 47.

In [71]:
ndf["age_at_transaction"] = (ndf["Today_Date"] - ndf["birthdate"]) / np.timedelta64(1, "Y")

In [76]:
ndf=ndf.drop('age_at_transaction',axis=1)

In [77]:
ndf.head()

cc_num provider    expires                 name sex  \
0  4796807885357879     visa 2023-05-01  Catherine Zimmerman   F   
1  4529266636192966     visa 2022-03-01     Michael Williams   M   
2  4922690008243953     visa 2027-02-01      Jessica Krueger   F   
3  4897369589533543     visa 2022-04-01          Ruth Harris   F   
4  4848518335893425     visa 2026-10-01          Paul Ashley   M   

                        mail  birthdate               City Country  AddedDate  \
0  valenciajason@hotmail.com 1988-09-20   Bryn Mawr-Skyway      US 2022-09-01   
1     brettkennedy@yahoo.com 1977-03-01  Gates-North Gates      US 2022-09-01   
2     marthacruz@hotmail.com 1947-09-10         Greenfield      US 2022-09-01   
3          james11@yahoo.com 1983-12-27           New City      US 2022-09-01   
4      matthew97@hotmail.com 1974-11-10            Peabody      US 2022-09-01   

  AddTime                 Today_Date  Age  days_until_card_expires  
0   19:50 2023-10-04 06:28:40.248442   35               -156.26991  
1   19:50 2023-10-04 06:28:40.248442   46               -582.26991  
2   19:50 2023-10-04 06:28:40.248442   76               1215.73009  
3   19:50 2023-10-04 06:28:40.248442   40               -551.26991  
4   19:50 2023-10-04 06:28:40.248442   49               1092.73009

In [84]:
#Their Visa card age

ndf=ndf.drop('Card_Age',axis=1)

ndf

cc_num provider    expires                 name sex  \
0     4796807885357879     visa 2023-05-01  Catherine Zimmerman   F   
1     4529266636192966     visa 2022-03-01     Michael Williams   M   
2     4922690008243953     visa 2027-02-01      Jessica Krueger   F   
3     4897369589533543     visa 2022-04-01          Ruth Harris   F   
4     4848518335893425     visa 2026-10-01          Paul Ashley   M   
...                ...      ...        ...                  ...  ..   
1995  4473593503484549     visa 2026-01-01        Megan Ramirez   F   
1996  4483607926610254     visa 2026-08-01        Jessica Gould   F   
1997  4249540384047310     visa 2022-10-01           Adam Black   M   
1998  4325889817240225     visa 2024-03-01      Brooke Robinson   F   
1999  4653925648407310     visa 2022-04-01      Mr. Jeff Torres   M   

                           mail  birthdate               City Country  \
0     valenciajason@hotmail.com 1988-09-20   Bryn Mawr-Skyway      US   
1        brettkennedy@yahoo.com 1977-03-01  Gates-North Gates      US   
2        marthacruz@hotmail.com 1947-09-10         Greenfield      US   
3             james11@yahoo.com 1983-12-27           New City      US   
4         matthew97@hotmail.com 1974-11-10            Peabody      US   
...                         ...        ...                ...     ...   
1995        jmorrison@gmail.com 1924-06-28        Stockbridge      US   
1996        larrydiaz@yahoo.com 1925-11-07      Port Richmond      US   
1997     richardrusso@yahoo.com 1980-09-15  Brentwood Estates      US   
1998        cwagner@hotmail.com 1986-11-05            Redding      US   
1999         robert76@gmail.com 1932-11-13             Tupelo      US   

      AddedDate AddTime                 Today_Date  Age  \
0    2022-09-01   19:50 2023-10-04 06:28:40.248442   35   
1    2022-09-01   19:50 2023-10-04 06:28:40.248442   46   
2    2022-09-01   19:50 2023-10-04 06:28:40.248442   76   
3    2022-09-01   19:50 2023-10-04 06:28:40.248442   40   
4    2022-09-01   19:50 2023-10-04 06:28:40.248442   49   
...         ...     ...                        ...  ...   
1995 2022-09-01   19:50 2023-10-04 06:28:40.248442   99   
1996 2022-09-01   19:50 2023-10-04 06:28:40.248442   98   
1997 2022-09-01   19:50 2023-10-04 06:28:40.248442   43   
1998 2022-09-01   19:50 2023-10-04 06:28:40.248442   37   
1999 2022-09-01   19:50 2023-10-04 06:28:40.248442   91   

      days_until_card_expires  
0                  -156.26991  
1                  -582.26991  
2                  1215.73009  
3                  -551.26991  
4                  1092.73009  
...                       ...  
1995                819.73009  
1996               1031.73009  
1997               -368.26991  
1998                148.73009  
1999               -551.26991  

[2000 rows x 14 columns]

In [85]:
ndf['City'].value_counts()

Newburgh          24
Estelle           20
Ada               19
Lino Lakes        18
Malden            18
                  ..
South Whittier     5
Nampa              5
Bon Air            5
Beaumont           4
Middle River       3
Name: City, Length: 161, dtype: int64

In [86]:
ndf['sex'].value_counts()

M    1010
F     990
Name: sex, dtype: int64

In [87]:
minmax('Age')

93   \
cc_num                             4659295496557788   
provider                                       visa   
expires                         2026-08-01 00:00:00   
name                                     Aaron Diaz   
sex                                               M   
mail                           morganneal@yahoo.com   
birthdate                       1921-09-03 00:00:00   
City                                        Bothell   
Country                                          US   
AddedDate                       2022-09-01 00:00:00   
AddTime                                       19:50   
Today_Date               2023-10-04 06:28:40.248442   
Age                                             102   
days_until_card_expires                  1031.73009   

                                                423  
cc_num                             4863447464793190  
provider                                       visa  
expires                         2024-01-01 00:00:00  
name                                 Jordan Mcgrath  
sex                                               F  
mail                      christianthomas@yahoo.com  
birthdate                       2004-01-01 00:00:00  
City                                       San Juan  
Country                                          US  
AddedDate                       2022-09-01 00:00:00  
AddTime                                       19:50  
Today_Date               2023-10-04 06:28:40.248442  
Age                                              19  
days_until_card_expires                    88.73009

102  Years Old is the Oldest. reside in Bothell city

19 years old is the Youngest reside in San juan city.